File: 2021/05 - Hydrothermal Venture/solution.js
/* --- Day 5: Hydrothermal Venture --- */

function silver() {
    let data = parseInput(x => x.match(/\d+/g).map(Number)),
        vents = data.filter(([x0, y0, x1, y1]) => x0 == x1 || y0 == y1);
    return countOverlaps(vents);
}

function gold() {
    let data = parseInput(x => x.match(/\d+/g).map(Number)),
        vents = data;
    return countOverlaps(vents);
}

function countOverlaps(vents) {
    let xMin = Math.min(...vents.map(([x0, y0, x1, y1]) => Math.min(x0, x1))),
        yMin = Math.min(...vents.map(([x0, y0, x1, y1]) => Math.min(y0, y1))),
        xMax = Math.max(...vents.map(([x0, y0, x1, y1]) => Math.max(x0, x1))),
        yMax = Math.max(...vents.map(([x0, y0, x1, y1]) => Math.max(y0, y1))),
        seafloor = Array.from(new Array(yMax - yMin + 1), () =>
            new Array(xMax - xMin + 1).fill(0)
        );
    vents.forEach(([x0, y0, x1, y1]) => {
        let xInc = Math.sign(x1 - x0),
            yInc = Math.sign(y1 - y0);
        while ((x1 - x0) * xInc >= 0 && (y1 - y0) * yInc >= 0) {
            seafloor[y0 - yMin][x0 - xMin]++;
            x0 += xInc;
            y0 += yInc;
        }
    });
    return sum(seafloor.map(row => row.filter(n => n > 1).length));
}

File: 2021/05 - Hydrothermal Venture/solution.py
#!/usr/bin/python3
# --- Day 5: Hydrothermal Venture ---

import re

# read input as list of coordinate pairs: x0, y0 -> x1, y1
input = open( "input.txt")
data = [ list(map( int, re.findall( r"\d+", x))) for x in input.readlines()]
input.close()

# helper functions
sign = lambda n: -1 if n < 0 else 1 if n > 0 else 0

def countOverlaps(vents):
    xMin = float("inf")
    xMax = -xMin
    yMin = xMin
    yMax = xMax
    for x0, y0, x1, y1 in vents:
        xMin = min( xMin, x0, x1)
        xMax = max( xMax, x0, x1)
        yMin = min( yMin, y0, y1)
        yMax = max( yMax, y0, y1)
    seafloor = [ [0] * (xMax - xMin + 1) for row in range(yMax - yMin + 1)]
    count = 0
    for x0, y0, x1, y1 in vents:
        xInc = sign(x1 - x0)
        yInc = sign(y1 - y0)
        while (x1 - x0) * xInc >= 0 and (y1 - y0) * yInc >= 0:
            if seafloor[y0 - yMin][x0 - xMin] == 1:
                count += 1
            seafloor[y0 - yMin][x0 - xMin] += 1
            x0 += xInc
            y0 += yInc
    return count

# solutions
def silver():
    vents = list(filter( lambda coords: coords[0] == coords[2] or coords[1] == coords[3], data))
    return countOverlaps(vents)

def gold():
    vents = data
    return countOverlaps(vents)
